openedx/features/course_experience/views/course_outline.py

```diff
@@ -23,15 +24,45 @@ def render_to_fragment(self, request, course_id=None, page_context=None, **kwarg
         """
         course_key = CourseKey.from_string(course_id)
         course_overview = get_course_overview_with_access(request.user, 'load', course_key, check_if_enrolled=True)
-
         course_block_tree = get_course_outline_block_tree(request, course_id)
+
         if not course_block_tree:
             return None
 
+        content_milestones = self.get_content_milestones(request, course_key)
+
         context = {
             'csrf': csrf(request)['csrf_token'],
             'course': course_overview,
             'blocks': course_block_tree,
+            'milestones': content_milestones
         }
         html = render_to_string('course_experience/course-outline-fragment.html', context)
         return Fragment(html)
+
+    def get_content_milestones(self, request, course_key):
+        """
+        Returns dict of subsections with prerequisites and whether the prerequisite has been completed or not
+        """
+
+        course_content_milestones = {}
+
+        all_course_prereqs = get_course_content_milestones(
+            course_id=course_key,
+            content_id=None,
+            relationship='requires',
+            user_id=None)
+
+        for milestone in all_course_prereqs:
+            course_content_milestones[milestone['content_id']] = {'completed_prereqs': True}
+
```

Comment:
Would be better to set the completed state of the milestone from within this loop to avoid having a second loop (starting on line 59) that gets the unfulfilled prereqs and then changes the value. Something like:
'''
for milestone in all_course_prereqs:
    course_content_milestones[milestone['content_id']] = {'completed_prereqs': milestones_api.user_has_milestone({'id': user_id}, milestone)}

Only problem with this is milestones_api.user_has_milestone() hits the database every time and I think milestone_helper.get_course_content_milestones() does some caching. Might make sense to expose a user_has_milestone() method in milestone_helper.py that does caching. Not sure if it's worth doing all of this, just a suggestion. Might want to ask Doug his take.

Reply:
in fact, if we did add user_has_milestone() to milestone_helper with caching my code could take advantage of it as well, as currently I'm directly calling milestones_api.user_has_milestone({'id': user_id}, milestone) to determine if prereq is met.

Reply:
Yeah I could add a 'user_has_milestone' function to the milestone_helper file.